### ISSUE TITLE
Output parameters are flagged as "Variable Foo is assigned but never read"

### ISSUE THREAD

**Issue by movedoa** [Riverside-Software/sonar-openedge#681]
Output parameters are flagged as "Variable Foo is assigned but never read" since 2.50.
I have about 3500 new issues because of this.
Afaik this was fixed in #25 in the old version.

![grafik](https://user-images.githubusercontent.com/13309951/58791812-2d981700-85f3-11e9-877c-654a6230faf7.png)

**Comment by stefandrissen** [comment]
Based on your code snippet I would say that the rule is correctly flagging it.

Where is lcErrorText read?

**Comment by movedoa** [comment]
It is never read but this was disabled on purpose in previous versions.
It produces a massive amount of issues and the only solution is to have a ton of overloads for these methods which just produces code for the sake of sonar.

I guess there are people who want this so maybe have output parameters as an option would be the best solution here.

**Comment by stefandrissen** [comment]
Disabling on purpose sounds like a custom rule.

You can use {&_proparse_ prolint-nowarn(varusage)} to suppress the rule on a specific variable (maybe that's what you mean with overload?).

**Comment by movedoa** [comment]
No i just mean polymorphism.

Doing `{&_proparse_ prolint-nowarn(varusage)}` is no solution for 3500 issues.
The rule should just have an option to flag output parameters or not which would solve my problem.

I am referencing #25 beacause it also solved this problem at the time.

> OUTPUT variables shouldn't be reported anymore as of version 1.0.7 of riverside-rules plugin

**Comment by stefandrissen** [comment]
> Doing `{&_proparse_ prolint-nowarn(varusage)}` is no solution for 3500 issues.
> The rule should just have an option to flag output parameters or not which would solve my problem.

IMHO decorating 3500 known cases is safer than disabling this rule for variables that are assigned but never read.

> I am referencing #25 because it also solved this problem at the time.
> 
> > OUTPUT variables shouldn't be reported anymore as of version 1.0.7 of riverside-rules plugin

That reads like it solved what you reported, being that output parameters should not be flagged when they are not read.

**Comment by movedoa** [comment]
Yeah it would be safer but also a bit ridiculous.
This is not a real issue in the code. I see no clean solution here to fix these issues exect for unnecessary polymorphism which produces slower code (method calls are very expensive).

...oh and i know it reads like it was another issue (which it was!) but that fix also fixed these cases afaik. Sorry if that wasn't clear.

**Comment by stefandrissen** [comment]
> Yeah it would be safer but also a bit ridiculous.

We have 59,406 proparse declarations (of which 57,433 generated) in our 4 millions lines of code (of which approx 3.5 million lines generated) - maybe we're a bit ridiculous. ;-)

> This is not a real issue in the code. I see no clean solution here to fix these issues exect for unnecessary polymorphism which produces slower code (method calls are very expensive).

I would never add run-time code to suppress issue. 
 
> ...oh and i know it reads like it was another issue (which it was!) but that fix also fixed these cases afaik. Sorry if that wasn't clear.

Aha, we will have to agree to differ on the meaning of "fixed".

**Comment by gquerret** [comment]
What has been changed in #25 is for this case:
```
procedure proc1:
  def output param prm1 as char no-undo.
  assign prm1 = "something".
end.
```
prm1 is not reported as assigned but never read. I'll double-check why this wasn't reported in the previous version.

**Comment by gquerret** [comment]
In your case, variable lcErrorText was also considered as read in the ANTLR2 parser. There's a change in the ANTLR4 parser, and lcErrorText is only considered as written (which is the correct behavior).

**Comment by gquerret** [comment]
Won't fix, the solution will be to do a bulk change in SonarQube to mark them as false positives.
